### ISSUE TITLE
PHPUnit data providers

### ISSUE THREAD

**Issue by devnix** [Innmind/Immutable#15]
I guess there is no way of implementing PHPUnit tests with data providers, right? 

Instead of this:
```php
public function testAddIsCommutative(): void
{
    $this->forAll(
        Set\Integers::any(),
        Set\Integers::any(),
    )->then(function (int $a, int $b) {
        self::assertSame(
            add($a, $b),
            add($b, $a),
        );
    });
}
```

We could also aim for this:
```php
#[DataProvider('additionProvider')]
public function testAddIsCommutative(int $a, int $b): void
{
    self::assertSame(
        add($a, $b),
        add($b, $a),
    );
}

public function additionProvider(int $a, int $b): iterable
{
    return $this->forAll(
        Set\Integers::any(),
        Set\Integers::any(),
    )->getIterator();
}
```

How would you feel about this approach? I think it allows to test more properties with the same generators with less code 😄

**Comment by Baptouuuu** [comment]
This could be added to `inmmind/black-box` 👍 however this means the shrinking mechanism can't be used.

**Comment by Baptouuuu** [comment]
[Shrinking](https://github.com/Innmind/BlackBox/blob/develop/documentation/proof.md#the-power-of-shrinking) is the process of _reducing_ the data that make the property fail. For example if a generated string contains a thousand characters but the property breaks above 255 characters then BlackBox will show you the string reduced to 255 characters and won't show you the one with a thousand.

However for this process to work BlackBox needs to be able to replay the property multiple times, this happens inside the `->then()` method call. When using a data provider it's PHPUnit that runs the test but BlackBox won't be aware if the test fail (and thus can't reduce the generated data).

**Comment by devnix** [comment]
Oh, I see, OK! 

About these two issues please excuse me, I wanted indeed to create them on the `inmind/black-box` repository but I guess I missed the tab 😅

Both issues can be moved to that repository if you want to 😄

**Comment by Baptouuuu** [comment]
No worries 😉 

No need to duplicate the issues. I'll try to work on these this weekend and comment here when it's done

**Comment by devnix** [comment]
Oh nice, thank you!

**Comment by Baptouuuu** [comment]
Released in `innmind/black-box:5.2.0` you can now do:

```php
#[DataProvider('additionProvider')]
public function testAddIsCommutative(int $a, int $b): void
{
    self::assertSame(
        add($a, $b),
        add($b, $a),
    );
}

public static function additionProvider(int $a, int $b): iterable
{
    return self::forAll(
        Set\Integers::any(),
        Set\Integers::any(),
    )->asDataProvider();
}
```

**Note**: `forAll` is called statically since PHPUnit 10 requires data providers to be static methods

**Comment by devnix** [comment]
Nice, thank you so much!
